js/compressed-token/src/program.ts

```diff
@@ -4,27 +4,137 @@ import {
+    ParsedTokenAccount,
+    TokenTransferOutputData,
+    accountCompressionProgram,
     bn,
     confirmConfig,
     defaultStaticAccountsStruct,
-    getConnection,
+    sumUpLamports,
     toArray,
     useWallet,
+    validateSameOwner,
+    validateSufficientBalance,
 } from '@lightprotocol/stateless.js';
 import {
     MINT_SIZE,
     TOKEN_PROGRAM_ID,
+    approve,
+    createApproveCheckedInstruction,
+    createApproveInstruction,
     createInitializeMint2Instruction,
 } from '@solana/spl-token';
 import { MINT_AUTHORITY_SEED, POOL_SEED } from './constants';
 import { Buffer } from 'buffer';
-import { TokenData } from './types';
+import { packCompressedTokenAccounts } from './instructions/pack-compressed-token-accounts';
+
+type CompressParams = {
+    /**
+     * The payer of the transaction.
+     */
+    payer: PublicKey;
+    /**
+     * owner of the *uncompressed* token account.
+     */
+    owner: PublicKey;
+    /**
+     * source (associated) token account address.
+     */
+    source: PublicKey;
+    /**
+     * owner of the compressed token account.
+     */
+    toAddress: PublicKey;
+    /**
+     * Mint address of the token to compress.
+     */
+    mint: PublicKey;
+    /**
+     * amount of tokens to compress.
+     */
+    amount: number | BN;
+    /**
+     * The state tree that the tx output should be inserted into.
+     */
+    outputStateTree: PublicKey;
+};
+
+type DecompressParams = {
+    /**
+     * The payer of the transaction.
+     */
+    payer: PublicKey;
+    /**
+     * input state to be consumed
+     */
+    inputCompressedTokenAccounts: ParsedTokenAccount[];
+    /**
+     * address of **uncompressed** destination token account.
+     */
+    toAddress: PublicKey;
+    /**
+     * amount of tokens to decompress.
+     */
+    amount: number | BN;
+    /**
+     * The state tree that the change tx output should be inserted into.
+     */
+    outputStateTree: PublicKey;
+    /**
+     * The recent state root indices of the input state. The expiry is tied to
+     * the proof.
+     */
+    recentInputStateRootIndices: number[];
+    /**
+     * The recent validity proof for state inclusion of the input state. It
+     * expires after n slots.
+     */
+    recentValidityProof: CompressedProof;
+};
+
+export type TransferParams = {
+    /**
+     * The payer of the transaction
+     */
+    payer: PublicKey;
+    /**
+     * The input state to be consumed
+     */
+    inputCompressedTokenAccounts: ParsedTokenAccount[];
+    /**
+     * Recipient address
+     */
+    toAddress: PublicKey;
+    /**
+     * Amount of tokens to transfer
+     */
+    amount: BN | number;
+    /**
+     * The recent state root indices of the input state. The expiry is tied to
+     * the proof.
+
+     */
+    recentInputStateRootIndices: number[];
+    /**
+     * The recent validity proof for state inclusion of the input state. It
+     * expires after n slots.
+     */
+    recentValidityProof: CompressedProof;
```

Comment:
nice the recent makes a lot of sense, we should add an error for validity proof expired!

Reply:
instead of the root not found error.

Reply:
yea we can add a method that explicitly validates the recency, but i think `root not found` is pretty decent for a start. (equivalent to `blockhash not found`); client can reason that it's most likely bc expired.